src/core/mesh/qgsmesh3daveraging.cpp

```diff
@@ -151,3 +248,513 @@ int QgsMeshSingleLevelAveragingMethod::verticalLevel() const
+  if ( startIndex >= 0 && startIndex < verticalLevels.size() )
+  {
+    startVerticalLevel = verticalLevels[ startIndex ];
+  }
+  else
+  {
+    startVerticalLevel = verticalLevels[ 0 ];
+  }
+
+  int endIndex = volumesBelowFaceCount - mStartVerticalLevel + 1;
+  if ( endIndex >= 0 && endIndex < verticalLevels.size() )
+  {
+    endVerticalLevel = verticalLevels[ endIndex ];
+  }
+}
+
+QgsMeshSigmaAveragingMethod::QgsMeshSigmaAveragingMethod()
+  : QgsMesh3dAveragingMethod( QgsMesh3dAveragingMethod::SigmaAveragingMethod )
+{
+}
+
+QgsMeshSigmaAveragingMethod::QgsMeshSigmaAveragingMethod( double startFraction, double endFraction )
+  : QgsMesh3dAveragingMethod( QgsMesh3dAveragingMethod::SigmaAveragingMethod )
+  , mStartFraction( startFraction )
+  , mEndFraction( endFraction )
+{
+  if ( mStartFraction > mEndFraction )
+  {
+    std::swap( mStartFraction, mEndFraction );
+  }
+}
+
+QgsMeshSigmaAveragingMethod::~QgsMeshSigmaAveragingMethod() = default;
+
+QDomElement QgsMeshSigmaAveragingMethod::writeXml( QDomDocument &doc ) const
+{
+  QDomElement elem = doc.createElement( QStringLiteral( "sigma-settings" ) );
+  elem.setAttribute( QStringLiteral( "start-fraction" ), startFraction() );
+  elem.setAttribute( QStringLiteral( "end-fraction" ), endFraction() );
+  return elem;
+}
+
+void QgsMeshSigmaAveragingMethod::readXml( const QDomElement &elem )
+{
+  mStartFraction = elem.attribute( QStringLiteral( "start-fraction" ) ).toInt();
+  mEndFraction = elem.attribute( QStringLiteral( "end-fraction" ) ).toInt();
+  if ( mStartFraction > mEndFraction )
+  {
+    std::swap( mStartFraction, mEndFraction );
+  }
+}
+
+bool QgsMeshSigmaAveragingMethod::equals( const QgsMesh3dAveragingMethod *other ) const
+{
+  if ( !other || other->method() != method() )
+    return false;
+
+  const QgsMeshSigmaAveragingMethod *otherMethod = static_cast<const QgsMeshSigmaAveragingMethod *>( other );
+
+  return qgsDoubleNear( otherMethod->startFraction(), startFraction() ) && qgsDoubleNear( otherMethod->endFraction(), endFraction() ) ;
+}
+
+QgsMesh3dAveragingMethod *QgsMeshSigmaAveragingMethod::clone() const
+{
+  return new QgsMeshSigmaAveragingMethod( startFraction(), endFraction() );
+}
+
+double QgsMeshSigmaAveragingMethod::startFraction() const
+{
+  return mStartFraction;
+}
+
+double QgsMeshSigmaAveragingMethod::endFraction() const
+{
+  return mEndFraction;
+}
+
+bool QgsMeshSigmaAveragingMethod::hasValidInputs() const
+{
+  return mStartFraction >= 0 && mEndFraction >= mStartFraction && mEndFraction <= 1;
+}
+
+void QgsMeshSigmaAveragingMethod::volumeRangeForFace(
+  double &startVerticalLevel,
+  double &endVerticalLevel,
+  const QVector<double> verticalLevels
+) const
+{
+  const double top = verticalLevels[ 0 ];
+  const double bot = verticalLevels[ verticalLevels.size() - 1 ];
+  const double diff = top - bot;
+
+  startVerticalLevel = bot + diff * mStartFraction;
+  endVerticalLevel = bot + diff * mEndFraction;
```

Comment:
Is this correct? What if vertical levels are not evenly distributed?

Reply:
distribution doesn't matter. startVerticalLevel/endVerticalLevel is in meters